aten/src/ATen/native/TensorShape.cpp

```diff
@@ -1335,81 +1335,487 @@ Tensor index_select_sparse(const Tensor& self, int64_t dim, const Tensor& index)
+          # indices at which to select values/indices.
+          # The CPP code uses binary search to find matches and may
+          # swap the loop order for better algorithmic complexity.
+          new_dim_indices = []
+          selected_dim_indices = []
+          # This is a brute-force algorithms to convey the main idea.
+          # The CPP code below is more efficient but more complicated.
+          for i, i_idx in enumerate(indices[dim]):
+              for j, j_idx in enumerate(index):
+                  if i_idx == j_idx:
+                      new_dim_indices.append(j)
+                      selected_dim_indices.append(i)
+          new_indices = indices.index_select(1, selected_dim_indices)
+          new_values = values.index_select(0, selected_dim_indices)
+          new_indices[dim] = new_dim_indices
       else:
           new_indices = indices
-          new_values[k] = values[k].index_select(dim - len(sparse_shape), index) for k in range(nnz)
+          new_values = values.index_select(dim - sparse_dim + 1, index);
     */
-  auto ndim = self.dim();
+  const auto ndim = self.dim();
   if (ndim == 0) {
     TORCH_CHECK_INDEX(false, "index_select() cannot be applied to a 0-dim tensor.");
   }
   if (!(index.dim() == 1 && index.dtype() == at::kLong)) {
     TORCH_CHECK_INDEX(false, "index_select() argument index must be 1-D long-tensor.");
   }
   dim = maybe_wrap_dim(dim, ndim);
-  auto size = self.size(dim);
-  auto sparse_dim = self.sparse_dim();
-  auto dense_dim = self.dense_dim();
-  auto indices = self._indices();
-  auto values = self._values();
-  auto nnz = values.size(0);
-  auto new_sizes = self.sizes().vec();
-  new_sizes[dim] = index.size(0);
+  const auto size = self.size(dim);
+  const auto sparse_dim = self.sparse_dim();
+  const auto dense_dim = self.dense_dim();
+  const auto indices = self._indices();
+  const auto values = self._values();
+  const auto nnz = values.size(0);
+  const auto index_len = index.size(0);
+  auto res_sizes = self.sizes().vec();
+  res_sizes[dim] = index_len;
+
+  // Equivalent to t.index_select(dim, idx), but vanilla index_select is not parallel,
+  // so we use gather instead.
+  // We use this method to select relevant indices/values
+  // from the intersection between indices[dim] and the index.
+  const auto index_select = [](const Tensor& t, int64_t dim, const Tensor& idx) -> Tensor {
+    const auto idx_len = idx.numel();
+    auto out_shape = t.sizes().vec();
+    out_shape[dim] = idx_len;
+    auto idx_shape = std::vector<int64_t>(t.dim(), 1);
+    idx_shape[dim] = idx_len;
+    return t.gather(dim, idx.view(idx_shape).expand(out_shape));
+  };
+
 
+  // If indexing into sparse dimensions
   if (dim < sparse_dim) {
+    const auto nneg_index = [&index, index_len, &self, size, dim](void) -> Tensor {
+      auto nneg_index = at::empty_like(index);
+      // nneg_index = ((index < 0) * index + size) + (index >= 0) * index
```

Comment:
It still remains wrong, I will fix that :) Thank you for pointing that out!

Reply:
I retract my words, the original version is correct.

Reply:
Take `size = 5` and `index = 3`. Then the original expression is
```
nneg_index = ((index < 0) * index + size) + (index >= 0) * index = ((0) * 3 + 5) + (1) * 3 = 8
```
which is obviously incorrect because then `nneg_index > size`. The actual code is correct, of course.